sensing/autoware_imu_corrector/src/gyro_bias_estimator.cpp

```diff
@@ -72,34 +102,186 @@ GyroBiasEstimator::GyroBiasEstimator(const rclcpp::NodeOptions & options)
     transform_listener_->get_latest_transform(imu_frame_, output_frame_);
   if (!tf_imu2base_ptr) {
-    RCLCPP_ERROR(
+    RCLCPP_ERROR_ONCE(
       this->get_logger(), "Please publish TF %s to %s", output_frame_.c_str(),
       (imu_frame_).c_str());
     return;
   }
 
   geometry_msgs::msg::Vector3Stamped gyro;
   gyro.header.stamp = imu_msg_ptr->header.stamp;
-  gyro.vector = transform_vector3(imu_msg_ptr->angular_velocity, *tf_imu2base_ptr);
+  if (scale_imu_.modify_imu_scale_) {
+    sensor_msgs::msg::Imu imu_msg = *imu_msg_ptr;
+    rclcpp::Time now_time = this->now();
+    auto imu_msg_mod = modify_imu(imu_msg, scale_imu_, now_time);
+    // Publish modified IMU
+    imu_scaled_pub_->publish(imu_msg_mod);
+    // Transform gyro from modified IMU
+    gyro.vector = transform_vector3(imu_msg_mod.angular_velocity, *tf_imu2base_ptr);
+  } else {
+    gyro.vector = transform_vector3(imu_msg_ptr->angular_velocity, *tf_imu2base_ptr);
+  }
+
+  gyro_all_.push_back(gyro);        // Used to update the gyro bias
+  gyro_scale_buf_.push_back(gyro);  // Used to update the scale
 
-  gyro_all_.push_back(gyro);
+  // Keep gyro buffer fixed size
+  if (gyro_scale_buf_.size() > static_cast<size_t>(buffer_size_gyro_)) {
+    gyro_scale_buf_.erase(gyro_scale_buf_.begin());
+  }
+
+  // Filter gyro data
+  gyro_yaw_rate_ = alpha_gyro_ * gyro_yaw_rate_ + (1 - alpha_gyro_) * gyro.vector.z;
+
+  // EKF always update p_
+  ekf_rate_.p_ = ekf_rate_.p_ + ekf_rate_.q_;
+
+  // Update diagnostics when necessary
+  if (gyro_info_.scale_status != diagnostic_msgs::msg::DiagnosticStatus::ERROR) {
+    if (ekf_rate_.p_ > ekf_rate_.max_variance_) {
+      gyro_info_.scale_summary_message =
+        "Covariance is high, scale estimation hasn't been updated for a while";
+      gyro_info_.scale_status = diagnostic_msgs::msg::DiagnosticStatus::WARN;
+      gyro_info_.scale_status_summary = "WARN";
+    }
+    if (ekf_rate_.p_ < ekf_rate_.min_covariance_) {
+      gyro_info_.scale_summary_message = "limiting covariance";
+      gyro_info_.scale_status = diagnostic_msgs::msg::DiagnosticStatus::WARN;
+      gyro_info_.scale_status_summary = "WARN";
+    }
+  }
+
+  // Limit covariance
+  if (ekf_rate_.p_ > ekf_rate_.max_variance_) {
+    ekf_rate_.p_ = ekf_rate_.max_variance_;
+  }
+  if (ekf_rate_.p_ < ekf_rate_.min_covariance_) {
+    ekf_rate_.p_ = ekf_rate_.min_covariance_;
+  }
 
   // Publish results for debugging
   if (gyro_bias_ != std::nullopt) {
+    // Angle is updated here but restarted when angle from pose is received
+    gyro_yaw_angle_ +=
+      (ekf_angle_.x_state_(1) * (gyro.vector.z) - gyro_bias_not_rotated_.value().z) * dt_imu;
+    ekf_angle_.x_state_(1) = (ekf_angle_.x_state_(1) * ekf_angle_.decay_coefficient_);
+
+    // EKF update
+    ekf_angle_.x_state_(0) = gyro_yaw_angle_;
+    Eigen::Matrix2d f_matrix;
+    f_matrix << 1, dt_imu * (gyro.vector.z - gyro_bias_not_rotated_.value().z), 0,
```

Comment:
Please make sure that `gyro_bias_not_rotated_` is checked for a value with `has_value()` before accessing `value()`.